pkg/services/ngalert/state/persister_async.go

```diff
@@ -54,6 +56,11 @@ func (a *AsyncStatePersister) fullSync(ctx context.Context, cache *cache) error
 	a.log.Debug("Full state sync start")
 	instances := cache.asInstances(a.doNotSaveNormalState)
 	if err := a.store.FullSync(ctx, instances); err != nil {
+		if errors.Is(err, store.ErrLockDB) {
+			a.log.Warn("Full state sync failed to acquire the lock, another full sync may be in progress")
+			return nil
```

Comment:
This feels potentially dangerous. What if there's a lock that somehow never gets unlocked (maybe an external factor we can't control, like perhaps someone left a manual session open on the grafana DB) - it'll just fail to save state in the background forever, and the only bread crumb is this log line.

What breaks if we return `err` here? If that's not viable, maybe we could retry a configurable number of times?

I realize this is bound to happen some in HA mode, but is there a way we can surface this a bit more? Maybe we could create a metric for it at a minimum to understand in cloud how often this might happen?

Reply:
> This feels potentially dangerous. What if there's a lock that somehow never gets unlocked (maybe an external factor we can't control, like perhaps someone left a manual session open on the grafana DB) - it'll just fail to save state in the background forever, and the only bread crumb is this log line

Yes, you are right, it is dangerous. Ideally, we'd want to use a transaction-level advisory lock instead, but unfortunately, MySQL doesn't support it. 

We could try to use [ServerLockService](https://github.com/grafana/grafana/blob/a813046f3d53767719637fe3e75d7030968f44d9/pkg/infra/serverlock/serverlock.go#L31) which [Matt mentioned](https://raintank-corp.slack.com/archives/C076RNRRZ2N/p1720533487109039?thread_ts=1720529260.275789&cid=C076RNRRZ2N). It won't have this issue because. However, this brings another problem because saving takes a lot of time, and the timeout should be even bigger.

> What breaks if we return err here? If that's not viable, maybe we could retry a configurable number of times?

If we return an error here, it will be just logged as an error and won't break anything, but I didn't want to log an error because failing to get the lock can happen in normal circumstances too. We can't detect if another Grafana holds the lock or it is just stuck and will never be unlocked. Sometimes, saving alert states takes minutes, so I think retries wouldn't help much either.

Reply:
I can add a metric with the timestamp of the last successful full sync. This way, it will be possible to monitor it and create alerts if needed. What do you think?

Reply:
[Added](https://github.com/grafana/grafana/pull/90034/commits/f7f52a87b68688d1228b01528b4a1698c6c4c77c) `grafana_alerting_state_full_sync_last_time`